### ISSUE TITLE
Check mapping exist

### ISSUE THREAD

**Issue by dengshuan** [olivere/elastic#97]
In get_mapping_test.go, there is a test to check whether a mapping exists. But we can't check it in our code since `buildURL` is not an exported method. So why not export this method?

**Comment by olivere** [comment]
The `buildURL` methods are internal because they are not part of the API that Elastic/Elasticsearch provides. Its the services that Elastic offers to the consumer.

There is no official API for checking whether a mapping exists or not. If you ask ES for a non-existing mapping, it returns a 404 with an error object (in 1.6):

```sh
$ curl -i -s -XGET 'localhost:9200/no-such-index/_mapping?pretty'
HTTP/1.1 404 Not Found
Content-Type: application/json; charset=UTF-8
Content-Length: 74

{
  "error" : "IndexMissingException[[no-such-index] missing]",
  "status" : 404
}
```

As Elastic currently treats `404 Not Found` not as an error but a valid response, it won't return an error in that case. Notice this might change with the next major version of Elastic which is in line with Elasticsearch 2.0.

In other words: If you want to find out whether a mapping exists, you could just check if the name of the mapping exists in the result from Elastic:

```go
mapping := "no-such-mapping"

res, err := client.GetMapping().Index(mapping).Do()
if err != nil {
  // Something went wrong
}
_, exists := res[mapping]
if !exists {
  // Mapping does not exist
}
```

I hope this solves your problem?!

**Comment by dengshuan** [comment]
Yes, thanks for your explanation, I just want to make sure a mapping exists before indexing a document, not to check the details of the mapping. So that solution is enough for me.

**Comment by olivere** [comment]
Okay. You probably should do that on startup or use something like [index templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html) to automatically create your indices if missing.

Anyway... closing this for now.
